### ISSUE TITLE
False positive in `extraction_operator_linter()` with `R6` class method calls

### ISSUE THREAD

**Issue by IndrajeetPatil** [r-lib/lintr#1485]
This shouldn't produce a lint:

``` r
library(lintr)

lint(
  text = "Person <- R6Class('Person', list(
          name = NULL,
          initialize = function(name) self$name <- name
          ))
  
  Bob <- Person$new('Bob')",
  linters = extraction_operator_linter()
)
#> <text>:6:16: warning: [extraction_operator_linter] Use `[[` instead of `$` to extract an element.
#>   Bob <- Person$new('Bob')
#>                ^
```

<sup>Created on 2022-07-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

And, although the suggested replacement would work here, this would be an extremely unusual way of calling `R6` class methods:

``` r
library(R6)

Person <- R6Class('Person', list(
  name = NULL,
  initialize = function(name) self$name <- name
))

Person[["new"]]('Bob')
#> <Person>
#>   Public:
#>     clone: function (deep = FALSE) 
#>     initialize: function (name) 
#>     name: Bob
```

<sup>Created on 2022-07-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

**Comment by MichaelChirico** [comment]
IMO this linter is just fundamentally problematic -- it's why we disabled it for {lintr}:

https://github.com/r-lib/lintr/blob/6d2b327b2cfbe0ba80d803e6a5245e47895825ab/.lintr#L17

If it were up to me I'd deprecate this linter. But if you do use it, then yes, the awkward `Person[["new"]]("Bob")` is indeed the correct usage. I'm not even sure how we'd parameterize our way out of this (just allow `$new`?)...

**Comment by IndrajeetPatil** [comment]
Is deprecating this linter an option?! Please let's do that then. 

This is problematic for R6, Shiny, and in so many other contexts. I haven't encountered a single codebase where I have run all linters and then decided to turn off extraction linter because it had produced so many false positives that attending to other, more relevant lints became a difficult task.

**Comment by MichaelChirico** [comment]
cc original author @fangly for whether we're missing something / he can see a good path forward for the linter. I'll also do a Mastodon post.

**Comment by IndrajeetPatil** [comment]
FWIW, maybe [codegrip]( https://github.com/lionel-/codegrip/issues/4) will provide a convenient way to do this conversion (`$` -> `[[`) where relevant.

**Comment by Bisaloo** [comment]
> Maybe allow all extractions that contain a call?

Another common awkward case that would be handled by excluding calls:

```r
knitr::opts_chunk$set(
  echo = TRUE
)
```

That said, I've disabled this linter in our organization as well because I find the `[[` syntax more difficult to read in many cases.
